PIPS/pip-1.md

```diff
@@ -0,0 +1,46 @@
+---
+pip: 1
+title: Permissionless LockProxies
+status: Draft
+author: John Wong (@jwheur)
+created: 2020-06-23
+---
+
+## Simple Summary
+The current LockProxy implementation requires an operator to control `setManagerProxy`, `bindProxyHash` and `bindAssetHash`.
+
+This PIP proposes an alternative implementation which does not require an operator.
+
+## Specification
+`setManagerProxy` can be restricted to be called once. After it is called the first time, the `managerProxyContract` address cannot be changed.
+
+`bindProxyHash` and `bindAssetHash` functions can be removed.
+
+The `lock` function can be modified to require the following parameters:
+- address fromAssetHash
+- uint64 toChainId
+- bytes memory targetProxyHash
+- bytes memory toAssetHash
+- bytes memory toAddress
+- uint256 amount
+
+The LockProxy contract should maintain a `balances` mapping of (bytes32 => uint256).
+This `balances` mapping should be updated within the `lock` function:
+```
+key = hash(fromAssetHash, toChainId, targetProxyHash, toAssetHash)
+balances[key] += amount
+```
+
+Where `hash` is any supported hashing function of the respective blockchain.
+
+The `unlock` function should be modified to require the following parameters:
+- bytes memory argsBs
+
+Where `argsBs` contains `fromAssetHash`, `fromChainId`, `fromProxyHash` and `toAssetHash`, `toAddress` and `amount`.
+
+The contract should check that the balances are sufficient, then reduce the balance and perform the unlock:
+```
+key = hash(toAssetHash, fromChainId, fromProxyHash, fromAssetHash)
+require(balances[key] >= amount)
+balances[key] -= amount
+```
```

Comment:
i think we can add notify info in this proposal

Reply:
sure, I will add it in

Reply:
thank you for the feedback!

Reply:
you're welcome

Reply:
This lockproxy can work at the "toChain" side, by which I mean, current interfaces of lockproxy seems fine at the Ethereum side for Neo to be crossed into Ethereum. However, they are not enough for the Neo asset to be crossed to ethereum at the Neo side. Say, if we deploy this contract in Neo chain and want to cross Neo into Ethereum, we would notice there is no way to map NeoAssetHashInEthereum with NeoAssetHashInNeo. Solving this, this proposal would be more compatible.

Reply:
Hmm I'm not sure what you mean, I think sending Neo to Ethereum should be ok:
1. On Ethereum's side, we deploy the Neo token contract 
2. The Neo token contract on Ethereum should mint the total supply to the Ethereum LockProxy and call `ethLockProxy.delegateAsset(4, <Neo LockProxy contract address>, c56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b, <Neo total supply>)`
3. To send Neo to Ethereum, a user would call `lock` on the Neo Lock Proxy: `neoLockProxy.lock(c56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b, 2, <Ethereum LockProxy contract address>, <Neo token contract hash from (1)>,  <to address>, <amount>)`

Reply:
it means in NEO side, user may be fill in a error mapping address, it doesn't check the parameters

Reply:
Oh I see, ok I understand now, I will make an update